### ISSUE TITLE
msa locale settings

### ISSUE THREAD

**Issue by sem-geologist** [hyperspy/hyperspy#1400]
I wanted to save my spectra to msa format however I got this:

```python
> max_pixel_spectrum
<EDSSEMSpectrum, title: Stack of EDX, dimensions: (|1347)>
> max_pixel_spectrum.save(filename='max_pix_spectra.msa',
                        extension='msa')

---------------------------------------------------------------------------
Error                                     Traceback (most recent call last)
<ipython-input-16-f43a23994200> in <module>()
      1 max_pixel_spectrum.save(filename='max_pix_spectra.msa',
----> 2                         extension='msa')

/mnt/bulk-data/development/git/hyperspy/hyperspy/signal.py in save(self, filename, overwrite, extension, **kwds)
   2027             basename, ext = os.path.splitext(filename)
   2028             filename = basename + '.' + extension
-> 2029         io.save(filename, self, overwrite=overwrite, **kwds)
   2030 
   2031     def _replot(self):

/mnt/bulk-data/development/git/hyperspy/hyperspy/io.py in save(filename, signal, overwrite, **kwds)
    464             overwrite = overwrite_method(filename)
    465         if overwrite is True:
--> 466             writer.file_writer(filename, signal, **kwds)
    467             _logger.info('The %s file was created' % filename)
    468             folder, filename = os.path.split(os.path.abspath(filename))

/mnt/bulk-data/development/git/hyperspy/hyperspy/io_plugins/msa.py in file_writer(filename, signal, format, separator, encoding)
    327             loc = locale.getlocale(locale.LC_TIME)
    328             if os_name == 'posix':
--> 329                 locale.setlocale(locale.LC_TIME, ('en_US', 'latin-1'))
    330             elif os_name == 'windows':
    331                 locale.setlocale(locale.LC_TIME, 'english')

/usr/lib/python3.5/locale.py in setlocale(category, locale)
    592         # convert to string
    593         locale = normalize(_build_localename(locale))
--> 594     return _setlocale(category, locale)
    595 
    596 def resetlocale(category=LC_ALL):

Error: unsupported locale setting
```

I am using debian testing. Why this is trying to use locale?

**Comment by francisco-dlp** [comment]
This format stores the time information using us locales, so they're required. We should document this better. In most OS they're installed by default, but in my Debian I had to install them manually too. Probably we should document this in the User Guide.

**Comment by sem-geologist** [comment]
We have in geology some crap geochronology (Isoplot and SQUID) software which uses 99.9% percent of geolochronologists, it works on specific versions of excel. But even better, then it requires that whole windows OS locale should be set to US, and it requires that printer should be installed before installing those plugins (No printer no plugin, even if 99.9% of use of those plogins is not printing). This is ridiculous, something really not well with plugin if it want to tweak os locales. It should not go to documentation, rather it should be fixed, and btw I think I have US locale.

**Comment by sem-geologist** [comment]
What will happen if I am doing something else at the same moment which depends from locale?

**Comment by francisco-dlp** [comment]
AFAIK you only need to have the US locale installed, you don't need to set it as default. I works for me in Debian too. Unfortunately we can't fix the specification. Is there any way to fix the code (i.e. make it work without using the OS locales)?

**Comment by sem-geologist** [comment]
I installed `locales-all`, and I am still getting same error...

**Comment by francisco-dlp** [comment]
I installed en_us utf8 locales using

```bash
sudo dpkg-reconfigure locales

```

and it works for me.

**Comment by sem-geologist** [comment]
I am talking about writer, it needs `latin-1`

**Comment by francisco-dlp** [comment]
That's the encoding.

**Comment by sem-geologist** [comment]
that is failing line in the msa pluggin

**Comment by sem-geologist** [comment]
it tries to switch os locale to the EN_US 'latin-1' for a moment, then after switches back.
IMHO this is overdoing, datetime library have enought of tweeks and whistles to get any datetime to any date and time formats thought about. I dont see the point in this locale switching

**Comment by francisco-dlp** [comment]
If that's possible, we should definitely fix it. However, this doesn't explain your error.

**Comment by sem-geologist** [comment]
In the linux systems in the recent years comes a lot of concurent stuff which tweeks everything what needs to be tweekd and does not need to be touched, dbus, systemd, even KDE5 does some ingerence to locales and times, I am really not sure what correct thing is to get this ok, but I am sure that there will be other distro/configuration in the wild where locale will fail. It is not good thing to change locale, and it is also possible that changing of locale on some systems can need root. What about BSD's?

**Comment by sem-geologist** [comment]
There is another can of worms: what is exactly the `metadata.General.date` format? would not be life 100% easier to have datetime there, that would simplify a things. We for example are doing FIB slicing at night, so first and last slice have different dates, it is impossible to sort them with time only. If that would be datetime, it would be very easy to export to anything.

**Comment by francisco-dlp** [comment]
From the [User Guide](http://hyperspy.org/hyperspy-doc/current/user_guide/metadata_structure.html):

> The acquisition or creation date in ISO 8601 date format

**Comment by vidartf** [comment]
This might be unrelated, but on Windows with Norwegian local, reading an MSA file will always fail on the first attempt. Subsequent attempts (for all files) will succeed. I think this is related to a Python issue, but not certain anymore.

**Comment by ericpre** [comment]
A better approach could be to use the babel library:
```python
from datetime import datetime
from babel.dates import format_date

d = datetime.today()
format_date(d, locale='en_US')
format_date(d, locale='en_GB')
```

**Comment by jat255** [comment]
I just ran into this on my Arch Linux system with a 3.7 conda environment:

```python
Error                                     Traceback (most recent call last)
<ipython-input-99-23f3709d6c34> in <module>
----> 1 s.sum((0,1)).save('calibrated_sum_spectrum.msa')

~/git_repos/hyperspy/hyperspy/signal.py in save(self, filename, overwrite, extension, **kwds)
   2144             basename, ext = os.path.splitext(filename)
   2145             filename = basename + '.' + extension
-> 2146         io.save(filename, self, overwrite=overwrite, **kwds)
   2147 
   2148     def _replot(self):

~/git_repos/hyperspy/hyperspy/io.py in save(filename, signal, overwrite, **kwds)
    512                              "False.")
    513         if write:
--> 514             writer.file_writer(filename, signal, **kwds)
    515             _logger.info('The %s file was created' % filename)
    516             folder, filename = os.path.split(os.path.abspath(filename))

~/git_repos/hyperspy/hyperspy/io_plugins/msa.py in file_writer(filename, signal, format, separator, encoding)
    326             loc = locale.getlocale(locale.LC_TIME)
    327             if os_name == 'posix':
--> 328                 locale.setlocale(locale.LC_TIME, ('en_US', 'latin-1'))
    329             elif os_name == 'windows':
    330                 locale.setlocale(locale.LC_TIME, 'english')

~/.conda/envs/hyperspy/lib/python3.7/locale.py in setlocale(category, locale)
    602         # convert to string
    603         locale = normalize(_build_localename(locale))
--> 604     return _setlocale(category, locale)
    605 
    606 def resetlocale(category=LC_ALL):

Error: unsupported locale setting
```

```sh
$ locale -a

C
en_US.utf8
POSIX
```

I generated the other `en_US` locale available on Arch by uncommenting `en_US ISO-8859-1` in `/etc/local.gen` and rerunning `# locale-gen`. After this, `locale -a` gives:

```
$ locale -a

C
en_US
en_US.iso88591
en_US.utf8
POSIX
```

 After restarting my kernel, I could save the file, but the date information was not written:

```
./git_repos/hyperspy/hyperspy/io_plugins/msa.py:339: UserWarning: I couldn't write the date information due toan unexpected error. Please report this error to the developers
  "I couldn't write the date information due to"
```

Upon further investigation, this is because there was no date/time information in the metadata (since I had loaded from a Numpy array). A agree with @sem-geologist that we should store date/time information as a proper `datetime`, and if there is not provided at signal creation, it should default to `now()` so a user doesn't get an error when saving like this.

As for the locale issue, @ericpre's solution sounds reasonable.

**Comment by ericpre** [comment]
This is an annoying issue, we should not change the locale... instead we should be using another library to do that, either pytz or babel. pytz is already a matploltib dependency and possibly other library so this should not be a big deal.

**Comment by ericpre** [comment]
Fixed in #2171.
